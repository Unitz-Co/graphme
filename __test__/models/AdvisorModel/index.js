import _ from 'lodash';
import graphme from '../../../index';

import Definition from '../Definition';

import GetAdvisor from '../../gql/advisor/GetAdvisor.gql';

import AdvisorProfileModel from '../AdvisorProfileModel';
import AdvisorPresenceModel from '../AdvisorPresenceModel';
import TransactionModel from '../TransactionModel';
import TransactionAggregateModel from '../TransactionAggregateModel';

class AdvisorModel extends graphme.BaseModel {
  static DEFINITION = Definition.create({
    name: 'Advisor',
    schema: {
      id: String,
      is_active: Boolean,
      created_at: Date,
    },
    nodes: [
      ['profile', 'AdvisorProfileModel', { column_mapping: { id: 'advisor_id' } }],
      ['presence', 'AdvisorPresenceModel', { column_mapping: { id: 'advisor_id' } }],
      ['transactions', 'TransactionModel.Collection', { column_mapping: { id: 'advisor_id' } }],
      [
        'transactions_aggregate',
        'TransactionAggregateModel',
        { column_mapping: { id: 'advisor_id' }, usePlanSync: true },
      ],
    ],
    key: 'id',

    baseQuery: GetAdvisor,

    GQL_ACTIONS: {
      FIND: `advisor`,
      GET: `advisor_by_pk`,
      INSERT: `insert_advisor_one`,
      UPDATE: `update_advisor_by_pk`,
      DELETE: `delete_advisor_by_pk`,
    },
  });
}

graphme.model('AdvisorModel', AdvisorModel);

export default AdvisorModel;
