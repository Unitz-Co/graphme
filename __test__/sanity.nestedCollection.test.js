import _ from 'lodash';
import AdvisorModel from './models/AdvisorModel';

const instance = AdvisorModel.fromData({ id: 'PdOJWFBgNPUEMhX1JlsDm7zWy012' });

test('nested collection getter', async () => {
  const transactions = await instance.transactions;
  console.snapshot('transactions', transactions.toObject());
});

test('nested collection getter', async () => {
  const transactions = await instance.transactions;
  await transactions.setArgs('limit: 2').sync();
  console.snapshot('transactions', transactions);

  await transactions.setArgs('limit: 2, order_by: {advisor_id: asc}').sync();
  console.snapshot('transactions', transactions);

  // await transactions.setArgs({ limit: 2, where: {id: { _in: ["123"] } } }).sync();
  // console.snapshot('transactions', transactions);

});

test('nested collection getter from find', async () => {
  const instances = await AdvisorModel.find({ where: {id: { _eq: 'PdOJWFBgNPUEMhX1JlsDm7zWy012'}} }, `
    id
    created_at
    profile {
      id
      display_name
    }
    transactions(limit: 2, order_by: {advisor_id: asc}) {
      id
      advisor_id
      user_id
      session_id
    }
  `);
  console.snapshot('instances', instances);
  const instance = _.first(instances);
  // const instnace = instances.getByPath('0');

  await instance.transactions.setArgs('limit: 2').sync();


  const transactions = await instance.transactions;
  await transactions.setArgs('limit: 2').sync();
  console.snapshot('transactions', transactions);

  await transactions.setArgs('limit: 2, order_by: {advisor_id: asc}').sync();
  console.snapshot('transactions', transactions);

  // await transactions.setArgs({ limit: 2, where: {id: { _in: ["123"] } } }).sync();
  // console.snapshot('transactions', transactions);

});

test('nested collection index access', async () => {
  const transactions = await instance.transactions;
  console.snapshot('transactions', transactions[0]);
  console.snapshot('advisor_id', await transactions[0].advisor_id);
  console.snapshot('user_id', await transactions[0].user_id);
  console.snapshot('session_id', await transactions[0].getByPath('session_id'));
});


test('nested collection mutations', async () => {
  const instance = AdvisorModel.fromData({ id: 'PdOJWFBgNPUEMhX1JlsDm7zWy012' });
  const transactions = await instance.transactions;

  const firstTran = _.first(transactions);
  const user_id = await firstTran.user_id;
  transactions.push({
    user_id,
    session_id: 'session_1',
  });

  transactions.push({
    user_id,
    session_id: 'session_1',
  });

  transactions.push({
    user_id,
    session_id: 'session_1',
  });

  transactions.push({
    user_id,
    session_id: 'session_last',
  });

  await instance.applyByPath('transactions.push', {
    user_id,
    session_id: 'session_last',
  });

  await transactions.save();

  console.snapshot('len', await transactions, transactions.length);

  const lastTran = _.last(transactions);
  console.snapshot('lastTran', await lastTran.session_id);


  const removedTrans = transactions.splice(10, 1000);

  console.snapshot('removedTrans.len', removedTrans.len);

  await Promise.all(removedTrans.map(item => item.delete()));

});


test('indirect nested model property getter', async () => {
  const profile = await instance.transactions;
  console.snapshot('profile.display_name (indirect)', await profile.display_name);
});

test('direct nested instance property getter', async () => {
  console.snapshot('profile.display_name (direct)', await instance.getByPath('profile.display_name'));
});

test('direct nested model property setter', async () => {
  const profile = await instance.profile;
  const oldName = await profile.display_name;

  console.snapshot('profile.display_name: (before)', oldName);
  const newName = `newVal_newName`;

  profile.display_name = newName;
  await profile.save();

  console.snapshot('profile.display_name: (change)', newName, await instance.getByPath('profile.display_name'));
  profile.display_name = oldName;
  await profile.save();

  console.snapshot('profile.display_name: (after)', await instance.getByPath('profile.display_name'));
});


test('direct nested model property setByPath', async () => {
  const oldName = await instance.getByPath('profile.display_name');
  console.snapshot('profile.display_name: (before)', oldName);
  const newName = `newVal_newName`;

  await instance.setByPath('profile.display_name', newName);
  await instance.applyByPath('profile.save');

  console.snapshot('profile.display_name: (change)', newName, await instance.getByPath('profile.display_name'));
  await instance.setByPath('profile.display_name', oldName);
  await instance.applyByPath('profile.save');

  console.snapshot('profile.display_name: (after)', await instance.getByPath('profile.display_name'));
});

test('direct nested model method call', async () => {
  const profile = await instance.profile;
  console.snapshot('getClass', profile.getClass());
});

test('direct nested model method applyByPath', async () => {
  console.snapshot('getClass', await instance.applyByPath('profile.getClass'));
});


