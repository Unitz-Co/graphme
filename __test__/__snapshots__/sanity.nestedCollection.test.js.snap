// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`direct nested instance property getter 1`] = `
Object {
  "args": Array [
    "profile.display_name (direct)",
    "new_displayName",
  ],
}
`;

exports[`direct nested model method applyByPath 1`] = `
Object {
  "args": Array [
    "getClass",
    [Function],
  ],
}
`;

exports[`direct nested model method call 1`] = `
Object {
  "args": Array [
    "getClass",
    [Function],
  ],
}
`;

exports[`direct nested model property setByPath 1`] = `
Object {
  "args": Array [
    "profile.display_name: (before)",
    "new_displayName",
  ],
}
`;

exports[`direct nested model property setByPath 2`] = `
Object {
  "args": Array [
    "profile.display_name: (change)",
    "newVal_newName",
    "newVal_newName",
  ],
}
`;

exports[`direct nested model property setByPath 3`] = `
Object {
  "args": Array [
    "profile.display_name: (after)",
    "new_displayName",
  ],
}
`;

exports[`direct nested model property setter 1`] = `
Object {
  "args": Array [
    "profile.display_name: (before)",
    "new_displayName",
  ],
}
`;

exports[`direct nested model property setter 2`] = `
Object {
  "args": Array [
    "profile.display_name: (change)",
    "newVal_newName",
    "newVal_newName",
  ],
}
`;

exports[`direct nested model property setter 3`] = `
Object {
  "args": Array [
    "profile.display_name: (after)",
    "new_displayName",
  ],
}
`;

exports[`indirect nested model property getter 1`] = `
Object {
  "args": Array [
    "profile.display_name (indirect)",
    undefined,
  ],
}
`;

exports[`nested collection getter 1`] = `
Object {
  "args": Array [
    "transactions",
    Array [
      Object {
        "id": "9fd256a1-ecfb-4458-a47d-77dd2ce6e01d",
      },
      Object {
        "id": "455e5e72-e0d5-4fa6-83a1-468c2d9322e3",
      },
      Object {
        "id": "a0231195-4b75-4359-9006-74ba6e289ee9",
      },
      Object {
        "id": "60fa64b4-c652-4fb6-b33a-63df1d20a21b",
      },
      Object {
        "id": "06f1a27e-7f5a-4d16-9ab7-c79ab3a3e99f",
      },
      Object {
        "id": "23a6ce73-eb10-4ece-a875-c3f3d0bc042e",
      },
      Object {
        "id": "56e52355-c117-471b-919a-d09dc7ca6aee",
      },
      Object {
        "id": "a682e5e0-c852-4b91-a5a4-9e368a446dac",
      },
      Object {
        "id": "b93c0877-b9e1-40d4-8cdd-427246a6e3ef",
      },
      Object {
        "id": "2a944866-4e4d-4c20-9dac-e45fa1a71de0",
      },
    ],
  ],
}
`;

exports[`nested collection getter 2`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection getter 3`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection getter 4`] = `
Object {
  "args": Array [
    "instances",
    CollectionMixins [
      AdvisorModel {},
    ],
  ],
}
`;

exports[`nested collection getter 5`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection getter 6`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection getter from find 1`] = `
Object {
  "args": Array [
    "instances",
    CollectionMixins [
      AdvisorModel {},
    ],
  ],
}
`;

exports[`nested collection getter from find 2`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection getter from find 3`] = `
Object {
  "args": Array [
    "transactions",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
    ],
  ],
}
`;

exports[`nested collection index access 1`] = `
Object {
  "args": Array [
    "transactions",
    TransactionModel {},
  ],
}
`;

exports[`nested collection index access 2`] = `
Object {
  "args": Array [
    "advisor_id",
    "PdOJWFBgNPUEMhX1JlsDm7zWy012",
  ],
}
`;

exports[`nested collection index access 3`] = `
Object {
  "args": Array [
    "user_id",
    "9KGhxtVGdPMD4VRIuZw25Iwo8x33",
  ],
}
`;

exports[`nested collection index access 4`] = `
Object {
  "args": Array [
    "session_id",
    "session_1",
  ],
}
`;

exports[`nested collection mutations 1`] = `
Object {
  "args": Array [
    "len",
    CollectionMixins [
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
      TransactionModel {},
    ],
    15,
  ],
}
`;

exports[`nested collection mutations 2`] = `
Object {
  "args": Array [
    "lastTran",
    "session_last",
  ],
}
`;

exports[`nested collection mutations 3`] = `
Object {
  "args": Array [
    "removedTrans.len",
    undefined,
  ],
}
`;
